File: py/bedrock/diff.py
"""Semantic diff (spec §8.1). Mirrors src/diff.ts: RFC6901 paths sorted
bytewise; additions/deletions use null; array differences report whole array."""
from __future__ import annotations

from .canon import canonical_string


def _escape(seg: str) -> str:
    return seg.replace("~", "~0").replace("/", "~1")


def _walk(prefix: str, before, after, out: list):
    if canonical_string(before) == canonical_string(after):
        return
    if isinstance(before, dict) and isinstance(after, dict):
        keys = set(before) | set(after)
        for k in sorted(keys):
            p = prefix + "/" + _escape(k)
            _walk(
                p,
                before[k] if k in before else None,
                after[k] if k in after else None,
                out,
            )
        return
    out.append({"path": prefix, "before": before, "after": after})


def semantic_diff(before, after) -> list:
    out: list = []
    _walk("", before, after, out)
    return sorted(out, key=lambda c: c["path"])

File: src/diff.ts
/**
 * Semantic diff (spec §8.1): compares two canonical objects; paths are RFC6901
 * JSON pointers sorted bytewise; additions/deletions use null for the absent
 * side; any array difference reports the whole array member.
 */
import { canonicalString, type Json } from "./canon.js";

export interface Change {
  path: string;
  before: Json;
  after: Json;
}

function escapePointer(seg: string): string {
  return seg.replace(/~/g, "~0").replace(/\//g, "~1");
}

function walk(prefix: string, before: Json, after: Json, out: Change[]): void {
  if (canonicalString(before) === canonicalString(after)) return;
  const bObj = isPlain(before);
  const aObj = isPlain(after);
  if (bObj && aObj) {
    const keys = new Set([...Object.keys(before as object), ...Object.keys(after as object)]);
    for (const k of [...keys].sort()) {
      const p = prefix + "/" + escapePointer(k);
      const b = (before as Record<string, Json>)[k];
      const a = (after as Record<string, Json>)[k];
      const bHas = Object.prototype.hasOwnProperty.call(before, k);
      const aHas = Object.prototype.hasOwnProperty.call(after, k);
      walk(p, bHas ? b! : null, aHas ? a! : null, out);
    }
    return;
  }
  // array differences report the whole array
  out.push({ path: prefix, before, after });
}

function isPlain(v: Json): boolean {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

export function semanticDiff(before: Json, after: Json): Change[] {
  const out: Change[] = [];
  walk("", before, after, out);
  return out.sort((x, y) => (x.path < y.path ? -1 : x.path > y.path ? 1 : 0));
}
